languages/ocaml.scm

```diff
@@ -1112,6 +1112,54 @@
   (#scope_id! "function_type")
 )
 
+; Allow softlines in infix expressions, such as
+; let b =
+;   foo
+;   || bar
+;   || baz
+
+; As above, infix expressions are nested grammar elements, so we must identify the
+; top-level one: it is the one that is not preceded by an infix operator.
+; We only consider the common logic operators, as not to mess with arithmetic expressions
+(
+  (infix_operator
+    [
+      "||"
+      "&&"
+    ]
```

Comment:
There are more than two infix operators in Ocaml (in fact there are infinitely many). Can we avoid having to list them all?

Reply:
This is deliberate, see the comment above: `We only consider the common logic operators, as not to mess with arithmetic expressions`.
There is a nice illustration in the test files:
```ocaml
let _ = len < 0 || srcoff < 0 || srcoff > src.position - len
  || dstoff < 0 || dstoff > (Bytes.length dst) - len
```
If we add a softline on every infix operator, it gets formatted as
```ocaml
let _ =
  len < 0
  || srcoff < 0
  || srcoff > src.position - len
  || dstoff < 0
  || dstoff
  > (Bytes.length dst)
  - len
```
Which is not the desired output.

Reply:
I see. But the question still stands. There are infinitely many operators in Ocaml, and we want to format them all. They have known priorities, can we use these priorities (or something in the syntax tree's structure) to format properly? (like each priority, or nesting level, or something should be a different scope, shouldn't it?).

Maybe this is too hard for 0.1. But this is definitely something that needs figuring out.

Reply:
I agree that each operator priority should define its own scope, ideally. However, I think we don't have access to the priorities, at the level at which we operate. We only have access to the syntax tree, which, for `a || b || c > d - e`, is
```
{ Node expression_item }
  { Node infix_expression }
    { Node value_path }
      { Node value_name }
    { Node infix_operator }
      { Node || }
    { Node infix_expression }
      { Node value_path }
        { Node value_name }
      { Node infix_operator }
        { Node || }
      { Node infix_expression }
        { Node value_path }
          { Node value_name }
        { Node infix_operator }
        { Node infix_expression }
          { Node value_path }
            { Node value_name }
          { Node infix_operator }
            { Node - }
          { Node value_path }
            { Node value_name }
```
The fact that the operator priorities are not exposed (or even known?) by the tree-sitter grammar is quite hindering.

Reply:
I thought it was just the priority of the first character in the symbol, which would be simpler for us (though not necessarily simple).

> We only have access to the syntax tree, which, for `a || b || c > d - e`, is

I guess that the best solution will look like making a scope per list of operators of the same name. We would have one scope for `||`, then one (nested) for `>` and one nested for `-`. We will probably want to build logic for this. Which is kind of what  #98 was proposing to do, I believe.